File: insight-brain-service/src/main/java/com/sonatype/insight/brain/git/SourceControlService.java
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
package com.sonatype.insight.brain.git;

import jakarta.inject.Inject;
import jakarta.inject.Named;
import jakarta.inject.Singleton;

import com.sonatype.insight.brain.dataaccess.sourcecontrol.SourceControlDAO;
import com.sonatype.insight.brain.dataaccess.sourcecontrol.SourceControlDefaultBranchCommitHistoryDAO;
import com.sonatype.insight.brain.dataaccess.sourcecontrol.SourceControlPullRequestCommentDAO;
import com.sonatype.insight.brain.model.sourcecontrol.SourceControl;
import com.sonatype.insight.brain.model.sourcecontrol.SourceControlEvent;
import com.sonatype.insight.brain.sourcecontrol.SourceControlUtils;
import com.sonatype.insight.dataaccess.TransactionContext;

@Named
@Singleton
public class SourceControlService
{
  private final SourceControlPullRequestCommentDAO sourceControlPullRequestCommentDAO;

  private final SourceControlDefaultBranchCommitHistoryDAO commitHistoryDAO;

  private final SourceControlDAO sourceControlDAO;

  private final SourceControlUtils sourceControlUtils;

  @Inject
  public SourceControlService(
      final SourceControlPullRequestCommentDAO sourceControlPullRequestCommentDAO,
      final SourceControlDefaultBranchCommitHistoryDAO commitHistoryDAO,
      final SourceControlDAO sourceControlDAO,
      final SourceControlUtils sourceControlUtils)
  {
    this.sourceControlPullRequestCommentDAO = sourceControlPullRequestCommentDAO;
    this.commitHistoryDAO = commitHistoryDAO;
    this.sourceControlDAO = sourceControlDAO;
    this.sourceControlUtils = sourceControlUtils;
  }

  public void onRepositoryUrlUpdated(SourceControlEvent sourceControlEvent) {
    try (TransactionContext tx = sourceControlDAO.createTransactionContext()) {
      tx.begin();
      SourceControl sourceControl = sourceControlDAO.getByOwnerId(sourceControlEvent.getApplicationId());
      if (sourceControl != null) {
        sourceControlDAO.updatePollTimeAndErrorCounts(tx, sourceControl.getId(), sourceControlEvent.getCreateTime(), 0);
        sourceControlPullRequestCommentDAO.deleteByApplicationId(tx, sourceControl.getOwnerId());
        commitHistoryDAO.deleteByApplicationId(tx, sourceControl.getOwnerId());
        sourceControlUtils.deleteCheckoutDirectory(sourceControlEvent.getApplicationId());
      }
      tx.commit();
    }
  }
}

File: insight-brain-frontend/src/main/frontend/owner.manager/source.control/source.control.service.js
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
import { prop } from 'ramda';

export default function SourceControlService($http, CLMLocations) {
  return {
    addSourceControlRecord: addSourceControlRecord,
    updateSourceControlRecord: updateSourceControlRecord,
    deleteSourceControlRecord: deleteSourceControlRecord,
    getProviderTypesMap: getProviderTypesMap,
    getProviderTypes: getProviderTypes,
    getCompositeSourceControlRecord: getCompositeSourceControlRecord,
    validateCompositeSCMConfig: validateCompositeSCMConfig,
    getSourceControlMetrics: getSourceControlMetrics,
  };

  /**
   * @param ownerId
   * @returns source control record with inherited fields
   */
  function getCompositeSourceControlRecord(ownerType, ownerId) {
    return $http.get(CLMLocations.getCompositeSourceControlUrl(ownerType, ownerId)).then(prop('data'));
  }

  function validateCompositeSCMConfig(ownerType, ownerId) {
    return $http.get(CLMLocations.getValidateScmConfigUrl(ownerType, ownerId)).then(prop('data'));
  }

  /**
   * @param ownerType   only applicable to APPLICATION
   * @param ownerId
   * @returns source control metrics associated with application
   */
  function getSourceControlMetrics(ownerType, ownerId) {
    return $http.get(CLMLocations.getSourceControlMetricsUrl(ownerType, ownerId)).then(prop('data'));
  }

  /**
   * Creates a source control record
   * @param ownerType Either application or organization
   * @param ownerId
   * @param sourceControl the source control object to be added
   * @returns source control record
   */
  function addSourceControlRecord(ownerType, ownerId, sourceControl) {
    let url = CLMLocations.getSourceControlUrl(ownerType, ownerId);
    let data = getDataFromSourceControl(ownerType, ownerId, sourceControl);
    return $http.post(url, data).then(prop('data'));
  }

  /**
   * Updates a source control record
   * @param ownerType Either application or organization
   * @param ownerId
   * @param sourceControl the source control object to be added
   * @returns source control record
   */
  function updateSourceControlRecord(ownerType, ownerId, sourceControl) {
    let url = CLMLocations.getSourceControlUrl(ownerType, ownerId);
    let data = getDataFromSourceControl(ownerType, ownerId, sourceControl);
    return $http.put(url, data).then(prop('data'));
  }

  /**
   * Deletes a source control record
   * @param ownerType Either application or organization
   * @param ownerId
   * @returns 204 on success
   */
  function deleteSourceControlRecord(ownerType, ownerId) {
    return $http.delete(CLMLocations.getSourceControlUrl(ownerType, ownerId)).then(prop('data'));
  }

  function getProviderTypesMap() {
    let providerTypesMap = {};
    getProviderTypes().forEach(function (providerType) {
      providerTypesMap[providerType.value] = providerType.name;
    });

    return providerTypesMap;
  }

  function getProviderTypes() {
    return [
      { name: 'Azure DevOps', value: 'azure' },
      { name: 'Bitbucket', value: 'bitbucket' },
      { name: 'GitHub', value: 'github' },
      { name: 'GitLab', value: 'gitlab' },
    ];
  }
}

function getDataFromSourceControl(ownerType, ownerId, sourceControl) {
  let data = {
    provider: sourceControl.provider,
    username: sourceControl.username,
    token: sourceControl.token,
    baseBranch: sourceControl.baseBranch,
    remediationPullRequestsEnabled: sourceControl.remediationPullRequestsEnabled,
    statusChecksEnabled: sourceControl.statusChecksEnabled,
    pullRequestCommentingEnabled: sourceControl.pullRequestCommentingEnabled,
    sourceControlScansEnabled: sourceControl.sourceControlScansEnabled,
    sourceControlScanTarget: sourceControl.sourceControlScanTarget,
  };
  if (ownerType === 'application') {
    data.repositoryUrl = sourceControl.repositoryUrl;
  }
  return data;
}

SourceControlService.$inject = ['$http', 'CLMLocations'];
